cloudweatherreport/run.py

```diff
@@ -213,49 +214,47 @@ def check_cloud_resource(self, test_plan, cloud_info):
         env = connect_juju_client(self.controller, logging=logging)
         if not env:
-            logging.error("Jujuclient could not connect to {} ".format(
-                self.controller))
+            msg = "Jujuclient could not connect to {} ".format(self.controller)
+            logging.error(msg)
+            test_result = self.generate_test_result(
+                provider=get_provider_name(
+                    guess_provider_name(self.controller)),
+                test_name="Juju client failed", output=msg)
+            self.save_result_in_datastore(test_result, [], test_plan)
             return False
+
         env.name = self.controller
         env_info = env.info()
         provider = (env_info.get("provider-type") or
                     env_info.get("ProviderType"))
         env.provider_name = get_provider_name(provider)
-
         logging.info('Running test on {}.'.format(env.provider_name))
         resource_available = self.check_cloud_resource(test_plan, env_info)
         if resource_available is False:
-            test_result = model.SuiteResult(
-                provider=env.provider_name,
-                test_outcome='INFRA',
-                tests=[model.TestResult(
-                    name='Resource not available',
-                    duration=0.0,
-                    output='Resource not available',
-                    result='INFRA',
-                    suite='Resource not available'
-                )])
-            benchmark_results = []
-        else:
-            try:
-                test_result = self.run_tests(test_plan, env)
-                benchmark_results = self.run_benchmarks(test_plan, env)
-            except Exception:
-                tb = traceback.format_exc()
-                error = "Exception ({}):\n{}".format(env.name, tb)
-                logging.error(error)
-                # create a fake SuiteResult to hold exception traceback
-                test_result = model.SuiteResult(
-                    provider=env.provider_name,
-                    test_outcome='Error Running Tests',
-                    tests=[model.TestResult(
-                        name='Error Running Tests',
-                        duration=0.0,
-                        output=error,
-                        result='INFRA',
-                    )])
-                return False
+            msg = 'Resource not available'
+            logging.error(msg)
+            test_result = self.generate_test_result(
+                provider=env.provider_name, test_name=msg, output=msg)
+            self.save_result_in_datastore(test_result, [], test_plan)
+            return False
 
+        benchmark_result = []
+        try:
+            test_result = self.run_tests(test_plan, env)
+            benchmark_result = self.run_benchmarks(test_plan, env)
+        except Exception:
+            tb = traceback.format_exc()
+            error = "Exception ({}):\n{}".format(env.name, tb)
+            logging.error(error)
```

Comment:
You can usually just do logging.exception here, and it will format the traceback for you.